csharp/extractor/Semmle.Extraction.CSharp/Extractor/Extractor.cs

```diff
@@ -321,73 +248,175 @@ private static IEnumerable<Action> ReadSyntaxTrees(IEnumerable<string> sources,
-                sw.Stop();
-                analyser.Logger.Log(Severity.Info, "  Extraction took {0}", sw.Elapsed);
+            var currentProcess = Process.GetCurrentProcess();
+            var cpuTime1 = currentProcess.TotalProcessorTime;
+            var userTime1 = currentProcess.UserProcessorTime;
 
-                foreach (var type in analyser.MissingNamespaces)
+            sw.Restart();
+            analyser.PerformExtraction(options.Threads);
+            sw.Stop();
+            var cpuTime2 = currentProcess.TotalProcessorTime;
+            var userTime2 = currentProcess.UserProcessorTime;
+
+            var performance = new Entities.PerformanceMetrics()
+            {
+                Frontend = new Entities.Timings() { Elapsed = elapsed, Cpu = cpuTime1, User = userTime1 },
+                Extractor = new Entities.Timings() { Elapsed = sw.Elapsed, Cpu = cpuTime2 - cpuTime1, User = userTime2 - userTime1 },
+                Total = new Entities.Timings() { Elapsed = stopwatch.Elapsed, Cpu = cpuTime2, User = userTime2 },
+                PeakWorkingSet = currentProcess.PeakWorkingSet64
+            };
+
+            logPerformance(performance);
+            analyser.Logger.Log(Severity.Info, "  Extraction took {0}", sw.Elapsed);
+
+            postProcess();
+
+            return analyser.TotalErrors == 0 ? ExitCode.Ok : ExitCode.Errors;
+        }
+
+        private static void AnalyseStandalone(
+            StandaloneAnalyser analyser,
+            IEnumerable<string> sources,
+            IEnumerable<string> referencePaths,
+            CommonOptions options,
+            IProgressMonitor progressMonitor,
+            Stopwatch stopwatch)
+        {
+            Analyse(stopwatch, analyser, options,
+                references => GetResolvedReferencesStandalone(referencePaths, references),
+                (analyser, syntaxTrees) => ReadSyntaxTrees(sources, analyser, null, null, syntaxTrees),
+                (syntaxTrees, references) => CSharpCompilation.Create("csharp.dll", syntaxTrees, references),
+                (compilation, options) => analyser.InitializeStandalone(compilation, options),
+                () => { },
+                _ => { },
+                () =>
                 {
-                    pm.MissingNamespace(type);
-                }
+                    foreach (var type in analyser.MissingNamespaces)
+                    {
+                        progressMonitor.MissingNamespace(type);
+                    }
+
+                    foreach (var type in analyser.MissingTypes)
+                    {
+                        progressMonitor.MissingType(type);
+                    }
 
-                foreach (var type in analyser.MissingTypes)
+                    progressMonitor.MissingSummary(analyser.MissingTypes.Count(), analyser.MissingNamespaces.Count());
+                });
+        }
+
+        private static ExitCode AnalyseNonStandalone(
```

Comment:
Should this be in `NonStandaloneExtractor.cs`?

Reply:
I didn't want to change even more things. I think the best would be to get rid of the `public static class Extractor` altogether and move its bits maybe to `StandaloneAnalyser` and `TracingAnalyser`.